internal/providers/vault/provider.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2022 Namespace Labs Inc; All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+
+package vault
+
+import (
+	"context"
+	"sync"
+
+	vaultclient "github.com/hashicorp/vault-client-go"
+	"namespacelabs.dev/foundation/framework/secrets/combined"
+
+	"github.com/hashicorp/vault-client-go/schema"
+	"namespacelabs.dev/foundation/internal/fnapi"
+	"namespacelabs.dev/foundation/internal/fnerrors"
+	"namespacelabs.dev/foundation/std/tasks"
+	"namespacelabs.dev/foundation/universe/vault"
+)
+
+func Register() {
+	p := &provider{
+		vaultClients: make(map[string]*vaultclient.Client),
+	}
+	combined.RegisterSecretsProvider(p.appRoleProvider)
+	combined.RegisterSecretsProvider(p.certificateProvider)
+}
+
+type provider struct {
+	mtx          sync.Mutex
+	vaultClients map[string]*vaultclient.Client
+}
+
+func (p *provider) Login(ctx context.Context, caCfg *vault.VaultProvider, audience string) (*vaultclient.Client, error) {
+	vKey := vaultIdentifier{
+		address:    caCfg.GetAddress(),
+		namespace:  caCfg.GetNamespace(),
+		authMethod: caCfg.GetAuthMethod(),
+	}
+	p.mtx.Lock()
```

Comment:
a Mutex is not cancelation aware. Which is why we rarely use mutexes (or we don't do expensive work under a lock).

We have a facility for this already, it's a `tcache`, you can use that instead.

Reply:
This too is copy/pasta, but let me learn about tcache.

Reply:
PTAL, there is some tcache/task nesting that I'm not sure of, but it seems to work (I see the task message showing up during login… it seems to work.)